docs/design/2022-11-17-ttl-table.md

```diff
@@ -0,0 +1,369 @@
+  - Whether to start TTL workers in the current instance or not.
+  - Scope: Instance
+  - Values: [ON, OFF]
+  - Default: ON
+
+### New Metrics
+
+We'll introduce some new metrics to monitor the TTL jobs:
+
+- `ttl_queries`
+  - The total count of select queries in TTL jobs
+  - Type: Counter
+  - Labels: type, table, result
+
+- `ttl_processed_expired_rows`
+  - The total count of expired rows processed in TTL jobs
+  - Type: Counter
+  - Labels: type, table
+
+- `ttl_query_duration`
+  - The duration of the queries in TTL jobs
+  - Labels: type
+  - Type: Histogram
+
+- `ttl_job_status`
+  - The status for the current TTL job. When the job is in the specified status, the value will 1, otherwise, it will be 0
+  - Labels: table, status
+  - Type: Gauge
+
+- `ttl_job_scan_workers`
+  - The running scan workers count for the ttl jobs
+  - Labels: table
+  - Type: Gauge
+
+In the above metrics, the optional values for type label is 'select' and 'delete' and the optional values for result label is `success` and `error`
+
+## Known Issues
+
+- Though the TTL jobs from different table runs distributively. However, one job from a table runs in a single TiDB node. If a table is very large, there may be a bottleneck.
+- Currently, the condition of generated column cannot be pushed down to TiKV. If a table uses generated column as a TTL time column, the filter will be performed in TiDB side. It brings some necessary network traffics and makes the query slow.
+
+## Future Works
+
+- Schedule the scan tasks from one jobs to different nodes. This work will take full advantage of the cluster resource, especially for big tables.
+- If an index with the prefix of the TTL time column exists, we can use it to query expire rows instead of scanning the full table. It will reduce the execution time of the scan tasks.  
+- Support TTL table as a parent table referred by a child table with 'ON DELETE CASCADE'. When some rows in a TTL table are deleted, the related rows in child table will be deleted too.
+- Support pushing down generated column condition to TiKV side. Or use the definition of generated column to construct condition directly.
+- Scan the table from TiFlash (if table has any TiFlash replica) instead of TiKV to reduce performance effect on TP business.
+- Dynamically adjust the runtime settings according to the current overhead of the cluster.
+- Add some builtin alters for cloud environment.
+
+## Alternative Solutions
+
+TiKV supports TTL on RawKV, so we may get question that whether we can implement TiDB TTL in same concept. IMO, this method has some downsides:
+  - Not SQL awareness.
+  - Customer may want to use any column as TTL Column, and may want to use Generated Column to convert other column type(JSON, varchar) to a DATETIME column as TTL Column.
+  - No future foreign key support.
+  - Not table awareness: TTL on RawKV is kv-level, can't easily set TTL on/off by Table
+  - Not compatible with CDC, Backup, Secondary Index
+
```

Comment:
You can see an earlier discussion: https://github.com/pingcap/tidb/pull/22763 (also documented), we also face some txn isolation problems. Leaving a choice to user seems to be a more flexible way, but it requires a user to be more "professional" because it not only affects performance, but also introduces some unexpected behaviors...

Reply:
Yeah I understand that would cause some issues and might limit users choices on CDC, isolation levels...
But we can have those documented and then allow users to make the appropriate choice. 
Because I believe in some scenarios, the scan and delete approach would be too expensive...

Reply:
I see, thanks for your suggestion @Yanhan0507 

We could treat this as a backlog: the most issue that concerned me is actually the conflicts with the snapshot isolation, but maybe we could add some 'read filters' to the storage to solve it. The whole solution would be complicated but possible.